x-pack/filebeat/module/suricata/eve/ingest/pipeline.json

```diff
@@ -19,83 +19,103 @@
       }
     }
   , { "set":
-      { "field": "ecs.event.type"
+      { "field": "event.type"
       , "value": "{{suricata.eve.event_type}}"
       }
     }
 
   , { "set":
-      { "field": "ecs.user_agent.raw"
+      { "field": "user_agent.original"
       , "value": "{{suricata.eve.http.http_user_agent}}"
       }
     }
   , { "user_agent":
-      { "field": "ecs.user_agent.raw"
-      , "target_field": "ecs.user_agent"
+      { "field": "user_agent.original"
+      , "target_field": "user_agent"
       , "ignore_missing": true
       }
     }
   , { "rename":
-      { "field": "ecs.user_agent.os"
-      , "target_field": "ecs.user_agent.temp_os"
+      { "field": "user_agent.os"
+      , "target_field": "user_agent.temp_os"
       , "ignore_missing": true
       }
     }
   , { "rename":
-      { "field": "ecs.user_agent.temp_os"
-      , "target_field": "ecs.user_agent.os.full_name"
+      { "field": "user_agent.temp_os"
+      , "target_field": "user_agent.os.full_name"
       , "ignore_missing": true
       }
     }
   , { "rename":
-      { "field": "ecs.user_agent.os_name"
-      , "target_field": "ecs.user_agent.os.name"
+      { "field": "user_agent.os_name"
+      , "target_field": "user_agent.os.name"
       , "ignore_missing": true
       }
     }
   , { "rename":
-      { "field": "ecs.user_agent.os_version"
-      , "target_field": "ecs.user_agent.os.version"
+      { "field": "user_agent.os_version"
+      , "target_field": "user_agent.os.version"
       , "ignore_missing": true
       }
     }
   , { "rename":
-      { "field": "ecs.user_agent.os_major"
-      , "target_field": "ecs.user_agent.os.major"
+      { "field": "user_agent.os_major"
+      , "target_field": "user_agent.os.major"
       , "ignore_missing": true
       }
     }
   , { "rename":
-      { "field": "ecs.user_agent.os_minor"
-      , "target_field": "ecs.user_agent.os.minor"
+      { "field": "user_agent.os_minor"
+      , "target_field": "user_agent.os.minor"
       , "ignore_missing": true
       }
     }
+  , { "script":
+      { "lang": "painless"
+      , "source": "if (ctx.suricata.eve.src_ip != null) { if (ctx.source_ecs == null) ctx.source_ecs = new HashMap(); ctx.source_ecs.ip = ctx.suricata.eve.src_ip; }"
+      }
+    }
```

Comment:
@webmat what do you think of this way to conditionally set a field only when the source field exists? I'm not familiar with ingest pipelines or painless

Reply:
Also, it looks like we have the option (not tested) of using the `convert` processor. Set type to be the original type and use ignore_missing and target_field. It's hacky but slightly cleaner than the above painless.

Reply:
There's a lie hidden in the name "Painless". I cannot imagine myself doing this for 50+ fields. And that's just Suricata...

I haven't looked at the `convert` processor. I'll go have a look.

But I do think we need a more declarative and compact way to do it. Imagine this was an Ingest Node plugin:

```JSON
{ "normalize":
  [ { "source": "suricata.eve.event_type", "target": "event.type", "error_missing": true },
    { "source": "suricata.eve.http.status", "target": "http.response.status_code", "conversion": "integer" },
    { ... }
  ]
},
```

By default `error_missing = false`, so the long list of fields we may have to convert would not need to specify that. But we could still force an error to be added to the event when really critical fields are missing, as I'm showing in the first event.

This example is loosely based on a plugin I was proposing we add to Logstash https://github.com/elastic/logstash/issues/9768 (for Suricata, when I started it as a LS module). Although I didn't keep pushing it when we moved Suricata to Beats. So it's still in the idea phase at this time for LS.

Reply:
Yeah convert could work, and it's not that hacky. But I feel that having one processor definition per field is pretty heavy. Plus the default is *not* to ignore missing, which makes the process heavier still. And since it's specifically tailored to convert, the `type` field is mandatory, when in this case I'd say 80%+ of the fields are straight renames that should remain as strings 😂

But I'd say let's go this way for now, it's the simplest approach, I think. The important part is to not have tons of empty fields :-)

```JSON
{ "convert":
  "field": "suricata.eve.event_type",
  "target_field": "event.type",
  "type": "string"
}, 
{ "convert":
  "field": "suricata.eve.http.status",
  "target_field": "http.response.status_code",
  "type": "integer",
  "ignore_missing": true
}, 
...
```

Reply:
Bad news, conversion doesn't seem to work:

```
Error loading pipeline: Error loading pipeline for fileset suricata/eve: couldn't load pipeline: couldn't load json. Error: 400 Bad Request: {"error":{"root_cause":[{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"},"suppressed":[{"type":"parse_exception","reason":"[type] type [ip] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [ip] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [text] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}}]}],"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"},"suppressed":[{"type":"parse_exception","reason":"[type] type [ip] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [ip] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [text] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}}]},"status":400}. Response body: {"error":{"root_cause":[{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"},"suppressed":[{"type":"parse_exception","reason":"[type] type [ip] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [ip] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [text] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}}]}],"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"},"suppressed":[{"type":"parse_exception","reason":"[type] type [ip] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [ip] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [text] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}},{"type":"parse_exception","reason":"[type] type [keyword] not supported, cannot convert field.","header":{"processor_type":"convert","property_name":"type"}}]},"status":400}```

Reply:
My mistake, I was using elasticsearch types when it expects JSON types. It works